File: solutions/scala/food-chain/2/src/main/scala/FoodChain.scala
sealed trait Animal:
  def name: String

private case class Fly(name: String) extends Animal

private class Rhyme(val name: String, val rhyme: String) extends Animal

private case class ExtraRhyme(override val name: String, override val rhyme: String) extends Rhyme(name, rhyme)

private case class Horse(name: String, deadLine: String) extends Animal

private val animals: Array[Animal] = Array(
  Fly("fly"),
  ExtraRhyme("spider","It wriggled and jiggled and tickled inside her."),
  Rhyme("bird","How absurd to swallow a bird!"),
  Rhyme("cat","Imagine that, to swallow a cat!"),
  Rhyme("dog","What a hog, to swallow a dog!"),
  Rhyme("goat","Just opened her throat and swallowed a goat!"),
  Rhyme("cow","I don't know how she swallowed a cow!"),
  Horse("horse","She's dead, of course!")
)

private val flyReason = "I don't know why she swallowed the fly. Perhaps she'll die."

private def openingLine(animal: Animal) = s"I know an old lady who swallowed a ${animal.name}."

private def optionalRhymeLine(animal: Animal): String = animal match {
  case rhyme: Rhyme => rhyme.rhyme
  case _ => ""
}

private def catchPhrase(swallower: Animal, caught: Animal): String = {
  val rhyme = caught match {
    case extra: ExtraRhyme => extra.rhyme.replace("It", " that")
    case _ => "."
  }
  s"She swallowed the ${swallower.name} to catch the ${caught.name}$rhyme"
}

private def catchLines(swallowedAnimals: Array[Animal]): List[String] = {
  swallowedAnimals
    .reverse
    .sliding(2)
    .collect { case Array(swallower, caught) => catchPhrase(swallower, caught) }
    .toList
}

object FoodChain {

  private def verse(number: Int): String = {
    val target = animals(number - 1)

    val body = target match {
      case Horse(_, deadLine) => List(deadLine)
      case _ => optionalRhymeLine(target) +: catchLines(animals.take(number)) :+ flyReason
    }

    (openingLine(target) +: body.filter(s => s.nonEmpty)).mkString("", "\n", "\n")
  }

  def recite(startAt: Int, endAt: Int): String = {
    (startAt to endAt)
      .map(verse)
      .mkString("", "\n", "\n")
  }


}

File: solutions/clojure/food-chain/1/src/food_chain.clj
(ns food-chain
  (:require [clojure.string :as str]))

(def animals [{:name "fly"}
              {:name "spider"
               :rhyme "It wriggled and jiggled and tickled inside her."
               :description "that wriggled and jiggled and tickled inside her"}
              {:name "bird" :rhyme "How absurd to swallow a bird!"}
              {:name "cat" :rhyme "Imagine that, to swallow a cat!"}
              {:name "dog" :rhyme "What a hog, to swallow a dog!"}
              {:name "goat" :rhyme "Just opened her throat and swallowed a goat!"}
              {:name "cow" :rhyme "I don't know how she swallowed a cow!"}
              {:name "horse" :rhyme "She's dead, of course!"}])

(def perhaps-die "I don't know why she swallowed the fly. Perhaps she'll die.")

(defn- old-lady-verse [animal]
  (format "I know an old lady who swallowed a %s." (animal :name)))

(defn- join-into-lines
  ([coll] (join-into-lines "\n" coll))
  ([sep coll] (str/join sep coll)))

(def double-sep "\n\n")

(defn- add-second-verse-rhyme [verses {:keys [rhyme]}]
  (cond-> verses
    rhyme (conj rhyme)))

(def why-did-she-swallowed (vec (partition 2 1 animals)))

(defn- she-swallowed-verse [n]
  (let [[catch swallowed] (get why-did-she-swallowed n)
        opt-description (if (:description catch) (str " " (:description catch)) "")]
    (format "She swallowed the %s to catch the %s%s." (:name swallowed) (:name catch) opt-description)))

(defn- add-previous-animals [verses start]
  (->> (range start)
       (map she-swallowed-verse)
       (into verses)))

(defn- not-horse-verses [verses start animal]
  (cond-> verses
    (not= "horse" (:name animal)) (-> (conj perhaps-die) (add-previous-animals start))))

(defn- verse [start]
  (let [animal (animals start)]
    (-> '()
        (not-horse-verses start animal)
        (add-second-verse-rhyme animal)
        (conj (old-lady-verse animal))
        join-into-lines)))

(defn recite
  "Returns the lyrics of the song: 'I Know an Old Lady Who Swallowed a Fly.'"
  [start-verse end-verse]
  (->> (range (dec start-verse) end-verse)
       (map verse)
       (join-into-lines double-sep)))


